docs/man/man3/MySqlConnector.3
.TH "MySqlConnector" 3 "Thu Sep 6 2018" "Version 1.0" "Contextualization Tool" \" -*- nroff -*-
.ad l
.nh
.SH NAME
MySqlConnector \- This is a class to access a MySQL database\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <mysqlconnector\&.h>\fP
.PP
Inherits \fBDatabaseConnectorAbstract\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBMySqlConnector\fP ()"
.br
.RI "Creates an empty \fBMySqlConnector\fP\&. "
.ti -1c
.RI "\fBMySqlConnector\fP (const QString host, const QString user, const QString password, const QString database=QString())"
.br
.RI "Creates a \fBMySqlConnector\fP setting values received by argument\&. "
.ti -1c
.RI "QString \fBgetUser\fP ()"
.br
.RI "Returns a user name to access database\&. "
.ti -1c
.RI "void \fBsetUser\fP (const QString user)"
.br
.RI "Sets username to access database\&. "
.ti -1c
.RI "void \fBsetPassword\fP (const QString password)"
.br
.RI "Sets password to access database\&. "
.ti -1c
.RI "QString \fBgetDataBase\fP ()"
.br
.RI "Returns a data base name to access database\&. "
.ti -1c
.RI "void \fBsetDataBase\fP (const QString dataBase)"
.br
.RI "Sets data base to access database\&. "
.ti -1c
.RI "QList< \fBString\fP * > \fBgetAllStrings\fP () override"
.br
.ti -1c
.RI "QList< \fBString\fP * > \fBgetStringsWithValue\fP (const QString &value, bool caseSensitive=true) override"
.br
.RI "Returns all strigns in database with the value received by argument\&. "
.ti -1c
.RI "QList< \fBString\fP * > \fBgetStringsWithApproximateValue\fP (const QString &value, bool caseSensitive=true) override"
.br
.RI "Returns all strigns in database with a similar value received by argument\&. "
.ti -1c
.RI "QList< \fBString\fP * > \fBgetStringWithId\fP (const QString &id, bool caseSensitive=true) override"
.br
.RI "Returns all strigns in database with the identifier received by argument\&. "
.ti -1c
.RI "QList< \fBString\fP * > \fBgetStringsWithState\fP (const QString state) override"
.br
.RI "Returns strings with state received by parameter\&. "
.ti -1c
.RI "bool \fBinsertString\fP (const \fBString\fP &string, const QString language) override"
.br
.RI "Inserts a new string in database\&. "
.ti -1c
.RI "int \fBinsertStrings\fP (const QList< \fBString\fP *> &strings, const QString language) override"
.br
.RI "Inserts a set of strings into the database\&. "
.ti -1c
.RI "int \fBremoveStringsWithValue\fP (const QString &value, bool caseSensitive=true) override"
.br
.RI "Removes from databse all strigns with the value received by argument\&. "
.ti -1c
.RI "int \fBremoveStringsWithId\fP (const QString &id, bool caseSensitive=true) override"
.br
.RI "Removes from databse the strign with the identifier received by argument\&. "
.ti -1c
.RI "QStringList \fBgetLanguages\fP () override"
.br
.RI "Returns all languages stored in database\&. "
.ti -1c
.RI "QStringList \fBgetLanguageIds\fP () override"
.br
.RI "Returns all language identifiers stored in database\&. "
.ti -1c
.RI "QString \fBgetKeyFromLanguage\fP (const QString language)"
.br
.RI "Returns the key of language received by parameter\&. "
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
This is a class to access a MySQL database\&. 
.PP
Definition at line 20 of file mysqlconnector\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "MySqlConnector::MySqlConnector (const QString host, const QString user, const QString password, const QString database = \fCQString()\fP)"

.PP
Creates a \fBMySqlConnector\fP setting values received by argument\&. 
.PP
\fBParameters:\fP
.RS 4
\fIuser\fP Username to access database\&. 
.br
\fIpassword\fP Password for the username\&. 
.br
\fIdatabase\fP Data base name\&. 
.br
\fItable\fP Table name\&. 
.RE
.PP

.PP
Definition at line 13 of file mysqlconnector\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "QList< \fBString\fP * > MySqlConnector::getAllStrings ()\fC [override]\fP, \fC [virtual]\fP"

.PP

.PP
Implements \fBDatabaseConnectorAbstract\fP\&.
.PP
Definition at line 64 of file mysqlconnector\&.cpp\&.
.SS "QString MySqlConnector::getDataBase ()"

.PP
Returns a data base name to access database\&. 
.PP
\fBReturns:\fP
.RS 4
User name\&. 
.RE
.PP

.PP
Definition at line 54 of file mysqlconnector\&.cpp\&.
.SS "QString MySqlConnector::getKeyFromLanguage (const QString language)"

.PP
Returns the key of language received by parameter\&. 
.PP
\fBParameters:\fP
.RS 4
\fIlanguage\fP Language to match key 
.RE
.PP
\fBReturns:\fP
.RS 4
Language ID 
.RE
.PP

.PP
Definition at line 326 of file mysqlconnector\&.cpp\&.
.SS "QStringList MySqlConnector::getLanguageIds ()\fC [override]\fP, \fC [virtual]\fP"

.PP
Returns all language identifiers stored in database\&. 
.PP
\fBReturns:\fP
.RS 4
Languages used\&. 
.RE
.PP

.PP
Implements \fBDatabaseConnectorAbstract\fP\&.
.PP
Definition at line 308 of file mysqlconnector\&.cpp\&.
.SS "QStringList MySqlConnector::getLanguages ()\fC [override]\fP, \fC [virtual]\fP"

.PP
Returns all languages stored in database\&. 
.PP
\fBReturns:\fP
.RS 4
Languages used\&. 
.RE
.PP

.PP
Implements \fBDatabaseConnectorAbstract\fP\&.
.PP
Definition at line 290 of file mysqlconnector\&.cpp\&.
.SS "QList< \fBString\fP * > MySqlConnector::getStringsWithApproximateValue (const QString & value, bool caseSensitive = \fCtrue\fP)\fC [override]\fP, \fC [virtual]\fP"

.PP
Returns all strigns in database with a similar value received by argument\&. 
.PP
\fBParameters:\fP
.RS 4
\fIvalue\fP \fBString\fP value\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
List with strings\&. 
.RE
.PP
If the value belongs to a string, the fwString is saved, otherwise relsease memory of fwString\&. If size of both strings is longer than MIN_LENGTH_FOR_APPROXIMATE and their size difference is less than MAX_LENGTH_DIFFERENCE, a value is considered valid if it is contained within the fwString value or vice versa\&. If size of any strings is less than MIN_LENGTH_FOR_APPROXIMATE, a value is considered valid only if it is equals than the value of fwString\&.
.PP
Implements \fBDatabaseConnectorAbstract\fP\&.
.PP
Definition at line 120 of file mysqlconnector\&.cpp\&.
.SS "QList< \fBString\fP * > MySqlConnector::getStringsWithState (const QString state)\fC [override]\fP, \fC [virtual]\fP"

.PP
Returns strings with state received by parameter\&. 
.PP
\fBParameters:\fP
.RS 4
\fIstate\fP State to be found\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
\fBString\fP list eith indicated state\&. 
.RE
.PP

.PP
Implements \fBDatabaseConnectorAbstract\fP\&.
.PP
Definition at line 185 of file mysqlconnector\&.cpp\&.
.SS "QList< \fBString\fP * > MySqlConnector::getStringsWithValue (const QString & value, bool caseSensitive = \fCtrue\fP)\fC [override]\fP, \fC [virtual]\fP"

.PP
Returns all strigns in database with the value received by argument\&. 
.PP
\fBParameters:\fP
.RS 4
\fIvalue\fP \fBString\fP value\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
List with strings\&. 
.RE
.PP

.PP
Implements \fBDatabaseConnectorAbstract\fP\&.
.PP
Definition at line 88 of file mysqlconnector\&.cpp\&.
.SS "QList< \fBString\fP * > MySqlConnector::getStringWithId (const QString & id, bool caseSensitive = \fCtrue\fP)\fC [override]\fP, \fC [virtual]\fP"

.PP
Returns all strigns in database with the identifier received by argument\&. 
.PP
\fBParameters:\fP
.RS 4
\fIvalue\fP \fBString\fP identifier\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
List with strings\&. 
.RE
.PP

.PP
Implements \fBDatabaseConnectorAbstract\fP\&.
.PP
Definition at line 153 of file mysqlconnector\&.cpp\&.
.SS "QString MySqlConnector::getUser ()"

.PP
Returns a user name to access database\&. 
.PP
\fBReturns:\fP
.RS 4
User name\&. 
.RE
.PP

.PP
Definition at line 39 of file mysqlconnector\&.cpp\&.
.SS "bool MySqlConnector::insertString (const \fBString\fP & string, const QString language)\fC [override]\fP, \fC [virtual]\fP"

.PP
Inserts a new string in database\&. Returns true if the insertion was succesfull, otherwise, returns false\&. 
.PP
\fBParameters:\fP
.RS 4
\fIstring\fP \fBString\fP instance to be inserted\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
bool 
.RE
.PP

.PP
Implements \fBDatabaseConnectorAbstract\fP\&.
.PP
Definition at line 212 of file mysqlconnector\&.cpp\&.
.SS "int MySqlConnector::insertStrings (const QList< \fBString\fP *> & strings, const QString language)\fC [override]\fP, \fC [virtual]\fP"

.PP
Inserts a set of strings into the database\&. Returns the number of inserted stirngs\&. 
.PP
\fBParameters:\fP
.RS 4
\fIstrings\fP List with strings to be added into databse\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
bool 
.RE
.PP

.PP
Implements \fBDatabaseConnectorAbstract\fP\&.
.PP
Definition at line 233 of file mysqlconnector\&.cpp\&.
.SS "int MySqlConnector::removeStringsWithId (const QString & id, bool caseSensitive = \fCtrue\fP)\fC [override]\fP, \fC [virtual]\fP"

.PP
Removes from databse the strign with the identifier received by argument\&. Returns the number of removed stirngs\&. 
.PP
\fBParameters:\fP
.RS 4
\fIvalue\fP \fBString\fP identifier\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
bool 
.RE
.PP

.PP
Implements \fBDatabaseConnectorAbstract\fP\&.
.PP
Definition at line 267 of file mysqlconnector\&.cpp\&.
.SS "int MySqlConnector::removeStringsWithValue (const QString & value, bool caseSensitive = \fCtrue\fP)\fC [override]\fP, \fC [virtual]\fP"

.PP
Removes from databse all strigns with the value received by argument\&. Returns the number of removed strings\&. 
.PP
\fBParameters:\fP
.RS 4
\fIvalue\fP \fBString\fP value\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
Number of removed strings 
.RE
.PP

.PP
Implements \fBDatabaseConnectorAbstract\fP\&.
.PP
Definition at line 244 of file mysqlconnector\&.cpp\&.
.SS "void MySqlConnector::setDataBase (const QString dataBase)"

.PP
Sets data base to access database\&. 
.PP
\fBParameters:\fP
.RS 4
\fIuser\fP User name\&. 
.RE
.PP

.PP
Definition at line 59 of file mysqlconnector\&.cpp\&.
.SS "void MySqlConnector::setPassword (const QString password)"

.PP
Sets password to access database\&. 
.PP
\fBParameters:\fP
.RS 4
\fIuser\fP Passwrod\&. 
.RE
.PP

.PP
Definition at line 49 of file mysqlconnector\&.cpp\&.
.SS "void MySqlConnector::setUser (const QString user)"

.PP
Sets username to access database\&. 
.PP
\fBParameters:\fP
.RS 4
\fIuser\fP User name\&. 
.RE
.PP

.PP
Definition at line 44 of file mysqlconnector\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Contextualization Tool from the source code\&.


src/storage/mysqlconnector.cpp
/**
 * @file mysqlconnector.cpp
 * @author Jorge Herrero Tardón (jorgeht@usal.es)
 * @date 20/02/2018
 * @version 1.0
 * @class MySqlConnector
 *
 * @brief This is a class to access a MySQL database.
 */

#include "mysqlconnector.h"

MySqlConnector::MySqlConnector(
    const QString host,
    const QString user,
    const QString password,
    const QString database
) : DatabaseConnectorAbstract() {
    QSqlDatabase db = QSqlDatabase::addDatabase("QMYSQL");
    bool ok;

    host_ = host;
    user_ = user;
    password_ = password;
    dataBase_ = database;

    db.setHostName(host_);
    db.setDatabaseName(dataBase_);
    db.setUserName(user_);
    db.setPassword(password_);

    ok = db.open();
    if (!ok) {
        Log::writeError(QString(Q_FUNC_INFO) + " MySQL connection failed");
        Log::writeError(QString(Q_FUNC_INFO) + " " + db.lastError().databaseText());
    }
}

QString MySqlConnector::getUser()
{
    return user_;
}

void MySqlConnector::setUser(const QString user)
{
    user_ = user;
}

void MySqlConnector::setPassword(const QString password)
{
    password_ = password;
}

QString MySqlConnector::getDataBase()
{
    return dataBase_;
}

void MySqlConnector::setDataBase(const QString dataBase)
{
    dataBase_ = dataBase;
}

QList<String *> MySqlConnector::getAllStrings()
{
    QSqlQuery query;
    QList<String *> out;
    bool ok;

    ok = query.exec("SELECT * FROM String");
    if (!ok) {
        Log::writeError(QString(Q_FUNC_INFO) + " " + query.lastError().databaseText());
    }

    while (query.next()) {
        out << new FirmwareString(
            query.value(0).toString(),
            query.value(1).toString(),
            query.value(2).toString(),
            query.value(3).toString(),
            query.value(4).toString()
        );
    }

    return out;
}

QList<String *> MySqlConnector::getStringsWithValue(const QString &value, bool caseSensitive)
{
    QSqlQuery query;
    QList<String *> out;
    bool ok;

    if(caseSensitive) {
        query.prepare("SELECT * FROM String WHERE upper(Value) = :value");
        query.bindValue(":value", value.toUpper());
    } else {
        query.prepare("SELECT * FROM String WHERE Value = :value");
        query.bindValue(":value", value);
    }

    ok = query.exec();
    if (!ok) {
        Log::writeError(QString(Q_FUNC_INFO) + " " + query.lastError().databaseText());
    }

    while (query.next()) {
        out << new FirmwareString(
            query.value(0).toString(),
            query.value(1).toString(),
            query.value(2).toString(),
            query.value(3).toString(),
            query.value(4).toString()
        );
    }

    return out;
}

QList<String *> MySqlConnector::getStringsWithApproximateValue(const QString &value, bool caseSensitive)
{
    QList<String *> out;
    Qt::CaseSensitivity isCaseSensitive = caseSensitive ? Qt::CaseSensitive : Qt::CaseInsensitive;

    foreach (String *string, getStringsWithValue(value, caseSensitive)) {
        /**
         * If the value belongs to a string, the fwString is saved, otherwise relsease memory of fwString.
         * If size of both strings is longer than MIN_LENGTH_FOR_APPROXIMATE and their size difference is less than
         * MAX_LENGTH_DIFFERENCE, a value is considered valid if it is contained within the fwString value or
         * vice versa.
         * If size of any strings is less than MIN_LENGTH_FOR_APPROXIMATE, a value is considered valid only if it
         * is equals than the value of fwString.
         */
        if (value.size() > MIN_LENGTH_FOR_APPROXIMATE && string->getValue().size() > MIN_LENGTH_FOR_APPROXIMATE &&
            qAbs(value.size() - string->getValue().size()) < MAX_LENGTH_DIFFERENCE) {
            if (value.contains(string->getValue(), isCaseSensitive) || string->getValue().contains(value, isCaseSensitive)) {
                out << string;
            } else {
                delete string;
            }
        } else {
            if (value.compare(string->getValue(), isCaseSensitive) == 0) {
                out << string;
            } else {
                delete string;
            }
        }
    }

    return out;
}

QList<String *> MySqlConnector::getStringWithId(const QString &id, bool caseSensitive)
{
    QSqlQuery query;
    QList<String *> out;
    bool ok;

    if(caseSensitive) {
        query.prepare("SELECT * FROM String WHERE upper(ID) = :id");
        query.bindValue(":value", id.toUpper());
    } else {
        query.prepare("SELECT * FROM String WHERE ID = :id");
        query.bindValue(":value", id);
    }

    ok = query.exec();
    if (!ok) {
        Log::writeError(QString(Q_FUNC_INFO) + " " + query.lastError().databaseText());
    }

    while (query.next()) {
        out << new FirmwareString(
            query.value(0).toString(),
            query.value(1).toString(),
            query.value(2).toString(),
            query.value(3).toString(),
            query.value(4).toString()
        );
    }

    return out;
}

QList<String *> MySqlConnector::getStringsWithState(const QString state)
{
    QSqlQuery query;
    QList<String *> out;
    bool ok;

    query.prepare("SELECT * FROM String WHERE State = :state");
    query.bindValue(":value", state);

    ok = query.exec();
    if (!ok) {
        Log::writeError(QString(Q_FUNC_INFO) + " " + query.lastError().databaseText());
    }

    while (query.next()) {
        out << new FirmwareString(
            query.value(0).toString(),
            query.value(1).toString(),
            query.value(2).toString(),
            query.value(3).toString(),
            query.value(4).toString()
        );
    }

    return out;
}

bool MySqlConnector::insertString(const String &string, const QString language)
{
    QSqlQuery query;

    query.prepare("INSERT INTO String (ID, Value, Description, MaxLenght, State, LanguageID) "
                  "VALUES (:id, :value, :description, :maxLength, :state, :languageId)");
    query.bindValue(":id", string.getId());
    query.bindValue(":value", string.getValue());
    query.bindValue(":description", string.getDescription());
    query.bindValue(":maxLength", string.getMaxLength());
    query.bindValue(":state", string.getState());
    query.bindValue(":languageId", getKeyFromLanguage(language));

    if (query.exec()) {
        Log::writeError(QString(Q_FUNC_INFO) + " " + query.lastError().databaseText());
        return true;
    }

    return false;
}

int MySqlConnector::insertStrings(const QList<String *> &strings, const QString language)
{
    int count = 0;

    foreach (String *string, strings) {
        count = insertString(*string, language) == true ? count+1 : count;
    }

    return count;
}

int MySqlConnector::removeStringsWithValue(const QString &value, bool caseSensitive)
{
    QSqlQuery query;
    bool ok;

    if(caseSensitive) {
        query.prepare("DELETE FROM String WHERE upper(Value) = :value");
        query.bindValue(":value", value.toUpper());
    } else {
        query.prepare("SELECT * FROM String WHERE Value = :value");
        query.bindValue(":value", value);
    }

    ok = query.exec();
    if (!ok) {
        Log::writeError(QString(Q_FUNC_INFO) + " " + query.lastError().databaseText());

        return 0;
    }

    return query.numRowsAffected();
}

int MySqlConnector::removeStringsWithId(const QString &id, bool caseSensitive)
{
    QSqlQuery query;
    bool ok;

    if(caseSensitive) {
        query.prepare("DELETE FROM String WHERE upper(ID) = :id");
        query.bindValue(":value", id.toUpper());
    } else {
        query.prepare("SELECT * FROM String WHERE ID = :id");
        query.bindValue(":value", id);
    }

    ok = query.exec();
    if (!ok) {
        Log::writeError(QString(Q_FUNC_INFO) + " " + query.lastError().databaseText());

        return 0;
    }

    return query.numRowsAffected();
}

QStringList MySqlConnector::getLanguages()
{
    QSqlQuery query;
    QStringList out;
    bool ok;

    ok = query.exec("SELECT Name FROM Language");
    if (!ok) {
        Log::writeError(QString(Q_FUNC_INFO) + " " + query.lastError().databaseText());
    }

    while (query.next()) {
        out << query.value(0).toString();
    }

    return out;
}

QStringList MySqlConnector::getLanguageIds()
{
    QSqlQuery query;
    QStringList out;
    bool ok;

    ok = query.exec("SELECT ID FROM Language");
    if (!ok) {
        Log::writeError(QString(Q_FUNC_INFO) + " " + query.lastError().databaseText());
    }

    while (query.next()) {
        out << query.value(0).toString();
    }

    return out;
}

QString MySqlConnector::getKeyFromLanguage(const QString language)
{
    QSqlQuery query;
    QStringList out;
    bool ok;

    query.prepare("SELECT ID FROM Language where Name = :language");
    query.bindValue(":language", language);

    ok = query.exec();
    if (!ok) {
        Log::writeError(QString(Q_FUNC_INFO) + " " + query.lastError().databaseText());
    }

    while (query.next()) {
        out << query.value(0).toString();
    }

    return out.isEmpty() ? QString() : out.first();
}

bool MySqlConnector::isAnyImportantFieldEmpty() {
    if (user_.isEmpty() || password_.isEmpty() || dataBase_.isEmpty()) {
        return true;
    }

    return false;
}
